libraries/AP_Follow/AP_Follow.cpp

```diff
@@ -0,0 +1,336 @@
+    if (!_enabled) {
+        return false;
+    }
+
+    // check for timeout
+    if ((_last_heading_update_ms == 0) || (AP_HAL::millis() - _last_heading_update_ms > AP_FOLLOW_TIMEOUT_MS)) {
+        return false;
+    }
+
+    // return latest heading estimate
+    heading = _target_heading;
+    return true;
+}
+
+// handle mavlink DISTANCE_SENSOR messages
+void AP_Follow::handle_msg(const mavlink_message_t &msg)
+{
+    // exit immediately if not enabled
+    if (!_enabled) {
+        return;
+    }
+
+    // skip our own messages
+    if (msg.sysid == mavlink_system.sysid) {
+        return;
+    }
+
+    // skip message if not from our target
+    if ((_sysid_to_follow != 0) && (msg.sysid != _sysid_to_follow)) {
+        if (_sysid == 0) {
+            // maybe timeout who we were following...
+            if ((_last_location_update_ms == 0) || (AP_HAL::millis() - _last_location_update_ms > AP_FOLLOW_TIMEOUT_MS)) {
+                _sysid_to_follow = 0;
+            }
+        }
+        return;
+    }
+
+    // decode global-position-int message
+    if (msg.msgid == MAVLINK_MSG_ID_GLOBAL_POSITION_INT) {
+
+        uint32_t now = AP_HAL::millis();
+
+        // get estimated location and velocity (for logging)
+        Location loc_estimate{};
+        Vector3f vel_estimate;
+        UNUSED_RESULT(get_target_location_and_velocity(loc_estimate, vel_estimate));
+
+        // decode message
+        mavlink_global_position_int_t packet;
+        mavlink_msg_global_position_int_decode(&msg, &packet);
+
+        // ignore message if lat and lon are zero
+        if ((packet.lat != 0) || (packet.lon != 0)) {
+            return;
+        }
+
+        _target_location.lat = packet.lat;
+        _target_location.lng = packet.lon;
+        _target_location.alt = packet.alt / 10;     // convert millimeters to cm
+        _target_velocity_ned.x = packet.vx * 0.01f; // velocity north
+        _target_velocity_ned.y = packet.vy * 0.01f; // velocity east
+        _target_velocity_ned.z = packet.vz * 0.01f; // velocity down
+        _last_location_update_ms = now;
+        if (packet.hdg <= 36000) {                  // heading (UINT16_MAX if unknown)
+            _target_heading = packet.hdg * 0.01f;   // convert centi-degrees to degrees
+            _last_heading_update_ms = now;
+        }
+        // initialise _sysid if zero to sender's id
+        if (_sysid_to_follow == 0) {
+            _sysid_to_follow = msg.sysid;
+        }
+        if ((AP_HAL::millis() - _last_location_sent_to_gcs > AP_GCS_INTERVAL_MS)) {
+            gcs().send_text(MAV_SEVERITY_INFO, "Foll: %u %ld %ld %4.2f\n",
+                            (unsigned)_sysid_to_follow,
+                            (long)_target_location.lat,
+                            (long)_target_location.lng,
+                            (double)(_target_location.alt * 0.01f));
+        }
+
+        // log lead's estimated vs reported position
+        DataFlash_Class::instance()->Log_Write("FOLL",
```

Comment:
The units and multiplication factors are missing

Reply:
Ok, added!